lib/client/api_matrix.es6.js

```diff
@@ -0,0 +1,129 @@
+        return this.ExpressionsSingleton.create();
+      },
+    },
+  ],
+  methods: [
+    {
+      name: 'toMatrix',
+      documentation: `An asynchronous method that takes an array of browser
+        keys, produce a nested matrix that contains browser and interface
+        relationship. This structure should be easy to be displayed as a
+        nested table in client side.`,
+      args: [
+        {
+          name: 'browserKeys',
+          typeName: 'StringArray',
+          documentation: `A string array contains valid browser keys. It will
+            be used for filtering browserAPIs.`,
+        },
+      ],
+      returns: {
+        typeName: 'JSON',
+        documentation: `Returns a JSON of the form:
+            {interfaceName: {apiName: {BrowserKey: true, ...} ...} ... } ...}
+            This structure is easy for displaying a nested table.`,
+      },
+      code: function(browserKeys) {
+        return this.browserAPIs
+          .where(this.M.IN(this.BrowserAPI.BROWSER_KEY, browserKeys))
+          .select(this.M.GROUP_BY( this.BrowserAPI.INTERFACE_KEY,
+            this.ArrayDAO.create()))
+          .then((groups) => {
+            let matrix = {};
+            let interfaceApis = groups.groupKeys;
+            for (let i = 0; i < interfaceApis.length; i++) {
+              let browserAPIs = groups.groups[interfaceApis[i]].array;
+              let interfaceName = browserAPIs[0].interfaceName;
+              let apiName = browserAPIs[0].apiName;
+              if (!matrix.hasOwnProperty(interfaceName)) {
+                matrix[interfaceName] = {};
+              }
+              matrix[interfaceName][apiName] = {};
+              for (let j = 0; j < browserAPIs.length; j++) {
+                matrix[interfaceName][apiName][
+                  browserAPIs[j].browserKey] = true;
+              }
+            }
+            return matrix;
+          });
+      },
+    },
+    {
+      name: 'toCSV',
+      documentation: `Takes an array of browser keys, produce a string of
+        CSV format.`,
+      args: [
+        {
+          name: 'browserKeys',
+          typeName: 'StringArray',
+          documentation: `A string array contains valid browser keys.`,
+        },
+      ],
+      returns: {
+        typeName: 'String',
+        documentation: `Returns a string of csv format.`,
+      },
+      code: function(browserKeys) {
+        return this.toMatrix(browserKeys).then(function(result) {
+          let table = [];
+          // Add table header.
+          table.push(['Interface', 'API'].concat(browserKeys));
+          let interfaces = Object.keys(result);
+          for (let i = 0; i < interfaces.length; i++) {
+            let interfaceName = interfaces[i];
+            let APIs = Object.keys(result[interfaceName]);
+            for (let j = 0; j < APIs.length; j++) {
+              let apiName = APIs[j];
+              let row = [interfaceName, apiName];
+              for (let browserKey in browserKeys) {
```

Comment:
@arobins do you mean that visually using `Object.keys()` + `for ( let k ... )` is more consistent, or do you mean that perhaps the structure produced by `toMatrix` should produce an array of objects that have `{ browserKey: ... }` in them?